[8.x] Add withColumn() to QueriesRelationships to support min, max, sum and…

khalilst opened the pull request:
I add a method named `withColumn` (copied/edited from `withCount`) method to support more SQL aggregation functions like `min, max, sum, avg` over relationships, e.g.
```php
Post::withCount('comments');
Post::withMax('comments', 'created_at');
Post::withMin('comments', 'created_at');
Post::withSum('comments', 'foo');
Post::withAvg('comments', 'foo');
```

Also, the `withCount` methods uses `withColumn` to avoid duplicated code and it works with backward-compatibility.
However, it is possible to call other/custom SQL function using `withColumn`, e.g.
```php
Post::withColumn('comments', 'created_at', 'distinct');
Post::withColumn('comments', 'content', 'length');
Post::withColumn('comments', 'created_at', 'custom_function');
```

In addition to, it is possible to add relationship's column to selected columns without SQL function, e.g.
```php
Comment::withColumn('post', 'title');
Post::withColumn('comments', 'content');
```
Both are working. Please consider the sub-query limited to have only *one* result.

khalilst committed:
Add withColumn() to QueriesRelationships to support min, max, sum and avg similar to withCount

driesvints commented:
This PR unfortunately contains breaking changes to a method signature. It also contains unrelated style changes to docblocks etc. Please send to master.

khalilst commented:
@driesvints 
I think there was a misunderstanding because of the similarity between `withCount()` and `withColumn()` source due to git diff.
The `withCount()` method has no changes to the method signature and it is completely the same as the original source.

```
    /**
     * Add subselect queries to count the relations.
     *
     * @param  mixed  $relations
     * @return $this
     */
    public function withCount($relations)
    {
        return $this->withColumn(is_array($relations) ? $relations : func_get_args(), '*', 'count');
    }
```

Just the body of the method has been changed.
And the changes to *docblocks* are completely *related*, because those blocks are *not* inside `withCount()` method, it is inside the general-purpose `withColumn()` method which is used by `withCount, withMin, winMax, ....`.

@taylorotwell 
Please review this PR again because of the mentiond misunderstandings.

khalilst commented:
@driesvints 
I think there was a misunderstanding because of the similarity between `withCount()` and `withColumn()` source due to git diff.
The `withCount()` method has no changes to the method signature and it is completely the same as the original source.

```php
    /**
     * Add subselect queries to count the relations.
     *
     * @param  mixed  $relations
     * @return $this
     */
    public function withCount($relations)
    {
        return $this->withColumn(is_array($relations) ? $relations : func_get_args(), '*', 'count');
    }
```

Just the body of the method has been changed.
And the changes to *docblocks* are completely *related*, because those blocks are *not* inside `withCount()` method, it is inside the general-purpose `withColumn()` method which is used by `withCount, withMin, winMax, ....`.

@taylorotwell 
Please review this PR again because of the mentiond misunderstandings.

driesvints commented:
@khalilst right, thanks for clarifying.

taylorotwell commented:
@khalilst did you follow this for this implementation? https://github.com/dwightwatson/aggregate/blob/master/src/AggregateServiceProvider.php

khalilst commented:
@taylorotwell 
Hmmm, honestly no, I did not know about that. However, seems there are similarities, *but*:
In fact, I was inspired from Builder's `sum, min, max, average` methods located in `src/Illuminate/Database/Query/Builder.php`.

I addition to those methods and support for other aggregation methods, I have add capability to return single column *without aggregation* which are not available in your given link.
Please consider the `limit 1`  in `withColumn`.

taylorotwell commented:
OK, after reviewing the implementation by @dwightwatson do you see any problems with this implementation?

khalilst commented:
@taylorotwell 
If you are asking me, I have to say no, there is no problem with this PR's implementation.
The `withCount()` automatic tests were successful.

taylorotwell commented:
@khalilst this PR forgot to add `loadAvg`, `loadMin`, etc. to Eloquent/Collection and Eloquent/Model. Note there are also methods for `loadMorphCount`.

khalilst commented:
@taylorotwell 
All right. I will add those methods to Eloquent/Collection, Eloquent/Model, and `loadMorth...`.

kanidjar commented:
hi @khalilst, @taylorotwell, the new withCount has broken our app.

We are using withCount inside a globalScope:

````php
Error: Maximum function nesting level of '256' reached, aborting!
``````

before this merge, everything was working correctly.

khalilst commented:
@kanidjar 
I will check it right now.

taylorotwell commented:
@khalilst let me know soon if you see the problem / fix - otherwise I will have to revert everything.

khalilst commented:
@taylorotwell 
I am already working on this. I will let you know the result.

khalilst commented:
@taylorotwell I tried different projects with the given code by @kanidjar and I have seen **no problem**.

@kanidjar You have edited your code several times. Are you sure there was not missing code here?
I have to metion that `withCount()` tests passed successfully.
Please paste the call-stack here.

kanidjar commented:
@khalilst here it is: https://gist.github.com/kanidjar/5594c42ced9b7494239bf28733f243a0

I've edited the code to remove all unrelated code.

A simple Item::all() with the global scope enabled is enough.

khalilst commented:
@kanidjar Thank you. I'm trying to find out the problem.

khalilst commented:
@taylorotwell @kanidjar I found the problem raised from line 385 at QueriesRelationships.php.
It was not in this PR, I think it was from code modifications/optimizations have done after merging by @sebastiaanluca.

I corrected this line and commit to open related PR #35029.

taylorotwell commented:
@khalilst we need the fix in a separate PR. Also, why does that cause the problem. The change @sebastiaanluca made is pretty valid that we would want to wrap the column names.

kanidjar commented:
@taylorotwell  I can confirm that applying the fix locally fixes the issue.

khalilst commented:
@taylorotwell Please take a look at the stack trace given by @kanidjar  [here](https://gist.github.com/kanidjar/5594c42ced9b7494239bf28733f243a0). 
You can see the exact line number 385 which the wrapping done here (at the bottom of gist).
I tested in my local environment and after modifying this line the reported issue resolved.
I will create another PR for this bugfix.

@kanidjar Yes, you are right.

taylorotwell commented:
We can call $this->getQuery()->getGrammar()->wrap() instead.

taylorotwell commented:
Pushed to 8.x. Please remove your fix from this from your other PR.

taylorotwell commented:
@khalilst can you confirm this commit fixes your problem if you apply the change directly in your vendor directory: https://github.com/laravel/framework/commit/20b0c6e19b635466f776502b3f1260c7c51b04ae

khalilst commented:
@taylorotwell Yes, it does.

I'm gonna remove the bugfix from other PR.

taylorotwell commented:
Released as 8.12.3.

robsontenorio commented:
@khalilst 

The `withCount` method allow us add constraints. How can we achieve that with `withMax`, `withMin` ?

```php
$posts = App\Models\Post::withCount(['votes', 'comments' => function (Builder $query) {
    $query->where('content', 'like', 'foo%');
}])->get();
```

khalilst commented:
@robsontenorio 
Same as `withCount`. Just consider to pass the second argument `$column`:

```php
Post::withMax(['votes', 'comments' => function (Builder $query) {
    $query->where('content', 'like', 'someting%');
}], 'column_name')->get();
```
**Note**: If you are using `with{Max, Min, Sum, Avg}` for an array of relations, the column should exist in all relations. 
In the above example, the `column_name` should be available in `votes & comments`,

robsontenorio commented:
@khalilst thank you!  The trick was second param `column_name` , not present in `withCount` (at least on laravel docs)

khalilst commented:
@robsontenorio You are welcome. I will create a PR to Laravel Doc asap.

Alexmg86 commented:
@taylorotwell 
Good afternoon!
I am a big fan of your work and in particular laravel is very upset. That year I tried to suggest you new things that were uploaded in this commit https://github.com/laravel/framework/pull/32474 but you rejected it and offered to write a package.
Which is what I did https://github.com/Alexmg86/laravel-sub-query
But this person just took all my work and added here and was accepted?
Very disappointing.

driesvints commented:
@Alexmg86 sometimes we reconsider feature requests. Don't be disappointed that your PR wasn't accepted at the time. Be happy that it's now finally in the core of the framework.

driesvints commented:
@Alexmg86 also, looking at your PR it wasn't closed because we didn't accept it. It was closed because it lacked a thorough explanation of the feature and tests.

Alexmg86 commented:
@driesvints But you could have said it, but a person just stole my work and this is at least bad. I'm really very upset

driesvints commented:
Can you point out which parts the maintainer stole from you?

Alexmg86 commented:
@driesvints Methods related to aggregation withMax, withMin, withSum, withAvg. You can see the dates of my commits, ideas.

driesvints commented:
Please link to them.

Alexmg86 commented:
@driesvints how to do it?

driesvints commented:
@Alexmg86 select the pieces of your code and paste them here so we can compare them.

khalilst commented:
> @taylorotwell
> Good afternoon!
> I am a big fan of your work and in particular laravel is very upset. That year I tried to suggest you new things that were uploaded in this commit #32474 but you rejected it and offered to write a package.
> Which is what I did https://github.com/Alexmg86/laravel-sub-query
> But this person just took all my work and added here and was accepted?
> Very disappointing.

@Alexmg86 I have **never** seen your code and I can not understand your insulting words without any proof.
You think you are the only person in this world that knows about aggregation.
I have just used Laravel's latest source **the old withCount** method and changed it to support other aggregation funcitons and single column as a result.

Alexmg86 commented:
@khalilst ok. I think so.

khalilst commented:
@Alexmg86 Sure, but definitely you were wrong

driesvints commented:
I'm going to close this off because I believe nothing more constructive can come out of this discussion.
